### DISCUSSION TITLE
How to open web link in default browser Android in a cross-platform project?

### DISCUSSION THREAD

**Discussion by GersonTommas** [AvaloniaUI/Avalonia#16983] [Q&A] [answered]
I've been trying to find the answer to this since two months now... I've tried a few of the answers which I can't remember all of them, but mostly...

`Process.Start("https://www.google.com")` in a Command on the view model attached to a button
`hyperlinkbutton` with `NavigationUri = "https://www.google.com"`
`ILauncher launcher = TopLevel.GetTopLevel(this).Launcher;` then `launcher.LaunchUriAsync(new System.Uri("https://www.google.com"));`

They all work just fine on windows... All of this... but none work on Android... I've found one single post that says to use "intent", but that is Avalonia.Android, which I don't have access from the main project... I CAN use intent from the .Android project itself... but how would I call it then? is there a way to create a main function in the main project, then overwrite it in each platform's project?

When I say it doesn't open on Android, I mean it doesn't do anything, I tap the button, it shows the animation indicating I pushed it, but that's it, no crash, no error, no nothing...

The idea is for the webpage to open in whatever default browser the user has... which again, All of the options I tried work fine if I launch on Windows Desktop, but they do nothing on Android...

I also found a lot of talk about this, but I was not able to understand the answers, as they never contained code, but they linked to different githubs with projects that I could not understand... most of them using some part of "MAUI.Essentials", which would be counter productive, as I intent this app to be also available for Linux, which I understand that MAUI does not support... So I would have to make a separated application just for linux...

I'm very sorry if the answer is out there, English is not my first language, and I've been searching for at least two months now (while developing other parts of the application)
Let me know if you need me to explain this better, or anything that I might be missing in my post...

**Comment by GersonTommas** [comment, marked as answer]
This was solved:
- Create an Interface (`IOpenWebPage`) in the main project
- Create classes on each project that implement that interface (AndroidOpenWebPage : IOpenWebPage --- DesktopOpenWebPage : IOpenWebPage)
- Set a static object in the shared project like `public static IOpenWebPage openWebPage;`(or however you called your Interface in step 1)
- In each project initialization process add `Shared.openWebPage = new AndroidOpenWebPage()` or `Shared.openWebPage = new DesktopWebPage()` use the one that corresponds to the platform. (`Shared` is the name of my class in the shared project, replace with whatever your class is called)

If anyone thinks an example would be useful I'll add one.

**Reply by thevortexcloud** [reply]
There is a better way. Avalonia now has the Launcher APIs which will do all of this for you.

https://docs.avaloniaui.net/docs/concepts/services/launcher

Previously you had to implement it yourself, which is why you found all those old posts about other ways of doing it.

**Reply by zipgenius** [reply]
Hello. I'm trying to use LaunchUriAsync in an Android project but it fails and it doesn't oipen the browser. I'm using Avalonia 11.5.x

**Reply by rabbitism** [reply]
> Hello. I'm trying to use LaunchUriAsync in an Android project but it fails and it doesn't oipen the browser. I'm using Avalonia 11.5.x

It should work. Iirc android has some limitation opening an non-ssl url. Can you try a https one first?

**Reply by zipgenius** [reply]
I tried directly with the https URL of my website and Chrome doesn't open. I tried both in the emulator and in my phone.
